# Data Cleaning: Ebay Germany Auto Ads 

## Purpose: 
* Develop understanding of what variables determine car prices.

## Goal: 
* Clean web scraped dataset by assessing raw dataset for errors, prepare for analysis.

----

## Dataset Description


* dateCrawled - When this ad was first crawled. All field-values are taken from this date.
* name - Name of the car.
* seller - Whether the seller is private or a dealer.
* offerType - The type of listing
* price - The price on the ad to sell the car.
* abtest - Whether the listing is included in an A/B test.
* vehicleType - The vehicle Type.
* yearOfRegistration - The year in which the car was first registered.
* gearbox - The transmission type.
* powerPS - The power of the car in PS.
* model - The car model name.
* kilometer - How many kilometers the car has driven.
* monthOfRegistration - The month in which the car was first registered.
* fuelType - What type of fuel the car uses.
* brand - The brand of the car.
* notRepairedDamage - If the car has a damage which is not yet repaired.
* dateCreated - The date on which the eBay listing was created.
* nrOfPictures - The number of pictures in the ad.
* postalCode - The postal code for the location of the vehicle.
* lastSeenOnline - When the crawler saw this ad last online.

----

## Source: 

----


In [1]:
import pandas as pd 
import numpy as np

In [2]:
#load csv into autos df
autos = pd.read_csv('/Users/jonathanmatsen/Documents/3_Projects/1_showcase_projects/data_cleaning_ebay_car_sales/autos.csv', encoding='Latin-1')

# > Explore
## >> _Dataset Overview_

### >>> _autos.head(1)_

In [3]:
#initial examination of df structure
autos.head(1)

dateCrawled                              name  seller offerType  \
0  2016-03-26 17:47:46  Peugeot_807_160_NAVTECH_ON_BOARD  privat   Angebot   

    price   abtest vehicleType  yearOfRegistration  gearbox  powerPS   model  \
0  $5,000  control         bus                2004  manuell      158  andere   

    odometer  monthOfRegistration fuelType    brand notRepairedDamage  \
0  150,000km                    3      lpg  peugeot              nein   

           dateCreated  nrOfPictures  postalCode             lastSeen  
0  2016-03-26 00:00:00             0       79588  2016-04-06 06:45:54

### >>> _autos.columns_

In [4]:
#overview of unique column names in df
columns1 = autos.columns

In [5]:
columns1

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
# create dictionary to map column indicies to snake case to enhance readability 
mapping_dict = {
    'dateCrawled': 'date_crawled',
     'name': 'name', 
     'seller': 'seller',
     'offerType': 'offer_type',
     'price': 'price', 
     'abtest': 'ab_test',
     'vehicleType': 'vehicle_type', 
     'yearOfRegistration': 'registration_year',
     'gearbox': 'gearbox', 
     'powerPS': 'power_ps',
     'model': 'model',
     'odometer': 'odometer',
     'monthOfRegistration': 'registration_month', 
     'fuelType': 'fuel_type', 
     'brand': 'brand',
     'notRepairedDamage': 'unrepaired_damage', 
     'dateCreated': 'ad_created', 
     'nrOfPictures': 'nr_of_pictures', 
     'postalCode': 'postal_code',
     'lastSeen': 'last_seen'
}  

In [7]:
#rename columns via mapping dict created above 
autos.rename(columns=mapping_dict, inplace=True)

In [8]:
#reexamine df to ensure column indicies correctly renamed 
autos.head(1)

date_crawled                              name  seller offer_type  \
0  2016-03-26 17:47:46  Peugeot_807_160_NAVTECH_ON_BOARD  privat    Angebot   

    price  ab_test vehicle_type  registration_year  gearbox  power_ps   model  \
0  $5,000  control          bus               2004  manuell       158  andere   

    odometer  registration_month fuel_type    brand unrepaired_damage  \
0  150,000km                   3       lpg  peugeot              nein   

            ad_created  nr_of_pictures  postal_code            last_seen  
0  2016-03-26 00:00:00               0        79588  2016-04-06 06:45:54

### >>> _autos.describe_

In [9]:
#initial examination of individual column statistics 
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-14 20:50:02  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

### >>> _autos.info_

In [10]:
#overview of column values and datatypes 
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
date_crawled          50000 non-null object
name                  50000 non-null object
seller                50000 non-null object
offer_type            50000 non-null object
price                 50000 non-null object
ab_test               50000 non-null object
vehicle_type          44905 non-null object
registration_year     50000 non-null int64
gearbox               47320 non-null object
power_ps              50000 non-null int64
model                 47242 non-null object
odometer              50000 non-null object
registration_month    50000 non-null int64
fuel_type             45518 non-null object
brand                 50000 non-null object
unrepaired_damage     40171 non-null object
ad_created            50000 non-null object
nr_of_pictures        50000 non-null int64
postal_code           50000 non-null int64
last_seen             50000 non-null object
dtypes: int64(5)


-----

## _Checkpoint: Findings_

#### _Columns to be reformatted_
* `odometer` - current type: object | desired type: int 
* `price` - current type: object | desired type: int 

#### _Potential column drop candidates_
* `seller` - only 2 unique values, with one of two unique values having a freq of 1
* `offer_type` - only 2 unique values, with one of two unique values having a freq of 1
* `nr_of_pictures` - not showing up on autos.describe

-----


# > Correct

## >> Review Reformat Candidate Columns

### >>> _'odometer.value_counts( )'_

In [11]:
#display odometer column value frequency to determine if all field values are of same structure 
#for purpose of reformatting column values from object -> int 
autos['odometer'].value_counts()

150,000km    32424
125,000km     5170
100,000km     2169
90,000km      1757
80,000km      1436
70,000km      1230
60,000km      1164
50,000km      1027
5,000km        967
40,000km       819
30,000km       789
20,000km       784
10,000km       264
Name: odometer, dtype: int64

### >>> _'price.value_counts( )'_

In [12]:
#display price column value frequency to determine if all field values are of same structure 
#for purpose of reformatting column values from object -> int 
autos['price'].value_counts(ascending=False).head()

$0        1421
$500       781
$1,500     734
$2,500     643
$1,000     639
Name: price, dtype: int64

## >> Review Drop Candidate Columns

### >>> _'seller.value_counts( )'_

In [13]:
# display seller column value frequency 
autos['seller'].value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

### >>> _'offer_type.value_counts( )'_

In [14]:
#display offer_type column value frequency
autos['offer_type'].value_counts()

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

### >>> _'nr_of_pictures'_

In [15]:
#display nr_of_pictures column value frequency
autos['nr_of_pictures']

0        0
1        0
2        0
3        0
4        0
        ..
49995    0
49996    0
49997    0
49998    0
49999    0
Name: nr_of_pictures, Length: 50000, dtype: int64

-----------------

### _Observations: Reformat & Drop_

#### _Reformat_

##### price 
* remove '$' and ',' 
* convert to int
      
##### odometer
* remove 'km' and ','
* convert to int

--------------------------

#### _Drop_

##### offer_type 
* drop -> _all values but one are identical -> meaningless for analysis_ 

##### seller 
* drop -> _all values but one are identical -> meaningless for analysis_ 

##### nr_of_pictures 
* drop -> _all values '0' -> meaningless for analysis_ 

------------------------



## >> Reformat 

### >>> _price_

In [16]:
# reformat price column
autos['price'] = autos['price'].str.replace('$', '').str.replace(',', '').astype(int)

### >>> _odometer_

In [17]:
# reformat odometer column
autos['odometer'] = autos['odometer'].str.replace('km', '').str.replace(',', '').astype(int)

In [18]:
# rename odometer column to odometer_km to capture relevant information deleted from field values during reformat from object to int
autos.rename(columns={'odometer': 'odometer_km'}, inplace=True)

## >> Drop
### >>> _'offer_type'_

In [19]:
# drop 'offer_type' column based on previous analysis 
autos = autos.drop('offer_type', axis=1)

### >>> _seller_

In [20]:
# drop 'seller' column based on previous analysis 
autos = autos.drop('seller', axis=1)

### >>> _'nr_of_pictures'_

In [21]:
# drop 'nr_of_pictures' column based on previous analysis 
autos = autos.drop('nr_of_pictures', axis=1)

## >> quality check 

In [22]:
#ensure reformatting changes correctly implemented 
autos.head(1)

date_crawled                              name  price  ab_test  \
0  2016-03-26 17:47:46  Peugeot_807_160_NAVTECH_ON_BOARD   5000  control   

  vehicle_type  registration_year  gearbox  power_ps   model  odometer_km  \
0          bus               2004  manuell       158  andere       150000   

   registration_month fuel_type    brand unrepaired_damage  \
0                   3       lpg  peugeot              nein   

            ad_created  postal_code            last_seen  
0  2016-03-26 00:00:00        79588  2016-04-06 06:45:54

# > Outlier and Suspect Values Analysis 

### Price

In [23]:
# price column min value
autos['price'].min()

0

In [24]:
# price column max value
autos['price'].max()

99999999

In [25]:
# number of unique values in price column 
autos['price'].unique().shape

(2357,)

In [26]:
# freq dist of price column
autos['price'].value_counts().sort_index(ascending=False).head(13)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
Name: price, dtype: int64

### Odometer_km

In [27]:
# odometer_km column min value
autos['odometer_km'].min()

5000

In [28]:
# odometer_km max value 
autos['odometer_km'].max()

150000

In [29]:
# number of unique values in odometer_km column 
autos['odometer_km'].unique().shape

(13,)

In [30]:
# freq dist of odometer_km column
autos['odometer_km'].value_counts().sort_index(ascending=False)

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
20000       784
10000       264
5000        967
Name: odometer_km, dtype: int64

# Findings: Outlier and Suspect Values 

### Price: 

* threshold to drop values = 999990
* 350,000, althought high, seems like at least a feasable number 
* whereas the next highest value is 999990 and every number above the\
* appears to be randomly keyed in or is possibly a fake ad 

### Odometer_km

* all values seem appropriate given the context 

------

# Remove Erroneous/Outlier Values from Price Column

In [31]:
# create boolean mask to filter rows to show price 350000 & below
autos = autos[(autos['price'] < 999990) & (autos['price'] != 0)]

In [32]:
# ensure that desired rows were removed based on price column decision criteria 
autos['price'].value_counts().sort_index().head(5)

1    156
2      3
3      1
5      2
8      1
Name: price, dtype: int64

In [33]:
# ensure that desired rows were removed based on price column decision criteria 
autos['price'].value_counts().sort_index(ascending=False).tail(5)

8      1
5      2
3      1
2      3
1    156
Name: price, dtype: int64

# >Checkpoint 

In [34]:
#review current state of df with changes made so far 
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 48565 entries, 0 to 49999
Data columns (total 17 columns):
date_crawled          48565 non-null object
name                  48565 non-null object
price                 48565 non-null int64
ab_test               48565 non-null object
vehicle_type          43979 non-null object
registration_year     48565 non-null int64
gearbox               46222 non-null object
power_ps              48565 non-null int64
model                 46107 non-null object
odometer_km           48565 non-null int64
registration_month    48565 non-null int64
fuel_type             44535 non-null object
brand                 48565 non-null object
unrepaired_damage     39464 non-null object
ad_created            48565 non-null object
postal_code           48565 non-null int64
last_seen             48565 non-null object
dtypes: int64(6), object(11)
memory usage: 6.7+ MB


In [35]:
# review current state of autos df after making desired changes 
autos.describe(include='all')

date_crawled         name          price ab_test vehicle_type  \
count                 48565        48565   48565.000000   48565        43979   
unique                46882        37470            NaN       2            8   
top     2016-03-29 23:42:13  Ford_Fiesta            NaN    test    limousine   
freq                      3           76            NaN   25019        12598   
mean                    NaN          NaN    5888.935591     NaN          NaN   
std                     NaN          NaN    9059.854754     NaN          NaN   
min                     NaN          NaN       1.000000     NaN          NaN   
25%                     NaN          NaN    1200.000000     NaN          NaN   
50%                     NaN          NaN    3000.000000     NaN          NaN   
75%                     NaN          NaN    7490.000000     NaN          NaN   
max                     NaN          NaN  350000.000000     NaN          NaN   

        registration_year  gearbox      power_ps  model    odometer_km  \
count        48565.000000    46222  48565.000000  46107   48565.000000   
unique                NaN        2           NaN    245            NaN   
top                   NaN  manuell           NaN   golf            NaN   
freq                  NaN    36102           NaN   3900            NaN   
mean          2004.755421      NaN    117.197158    NaN  125770.101925   
std             88.643887      NaN    200.649618    NaN   39788.636804   
min           1000.000000      NaN      0.000000    NaN    5000.000000   
25%           1999.000000      NaN     71.000000    NaN  125000.000000   
50%           2004.000000      NaN    107.000000    NaN  150000.000000   
75%           2008.000000      NaN    150.000000    NaN  150000.000000   
max           9999.000000      NaN  17700.000000    NaN  150000.000000   

        registration_month fuel_type       brand unrepaired_damage  \
count         48565.000000     44535       48565             39464   
unique                 NaN         7          40                 2   
top                    NaN    benzin  volkswagen              nein   
freq                   NaN     29368       10336             34775   
mean              5.782251       NaN         NaN               NaN   
std               3.685595       NaN         NaN               NaN   
min               0.000000       NaN         NaN               NaN   
25%               3.000000       NaN         NaN               NaN   
50%               6.000000       NaN         NaN               NaN   
75%               9.000000       NaN         NaN               NaN   
max              12.000000       NaN         NaN               NaN   

                 ad_created   postal_code            last_seen  
count                 48565  48565.000000                48565  
unique                   76           NaN                38474  
top     2016-04-03 00:00:00           NaN  2016-04-07 06:17:27  
freq                   1887           NaN                    8  
mean                    NaN  50975.745207                  NaN  
std                     NaN  25746.968398                  NaN  
min                     NaN   1067.000000                  NaN  
25%                     NaN  30657.000000                  NaN  
50%                     NaN  49716.000000                  NaN  
75%                     NaN  71665.000000                  NaN  
max                     NaN  99998.000000                  NaN

In [36]:
# count of unique column values 
autos.columns.nunique()

17


-----


# Analysis: Date/Time Columns 

### date_crawled freq as % of Total (by day) 

In [37]:
# date_crawled frequency percentages 

crawl_date_freq = autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False, ascending=False).sort_index
crawl_date_freq

<bound method Series.sort_index of 2016-04-03    0.038608
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-12    0.036920
2016-03-14    0.036549
2016-04-04    0.036487
2016-03-07    0.036014
2016-04-02    0.035478
2016-03-28    0.034860
2016-03-19    0.034778
2016-03-15    0.034284
2016-03-29    0.034099
2016-04-01    0.033687
2016-03-30    0.033687
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-22    0.032987
2016-03-11    0.032575
2016-03-23    0.032225
2016-03-26    0.032204
2016-03-10    0.032184
2016-03-31    0.031834
2016-03-17    0.031628
2016-03-25    0.031607
2016-03-27    0.031092
2016-03-16    0.029610
2016-03-24    0.029342
2016-03-05    0.025327
2016-03-13    0.015670
2016-03-06    0.014043
2016-04-05    0.013096
2016-03-18    0.012911
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64>

### ad_created freq as % of Total (by day)

In [38]:
# ad_created frequency percentages 
ad_created_freq = autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False, ascending=False).sort_index()
ad_created_freq

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64

### last_seen freq as % of Total (by day)

In [39]:
# last_seen frequency percentages 

last_seen_freq = autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False, ascending=False).sort_index()
last_seen_freq

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64


-----


## registration_year: a closer look 

`Moving along, the registration_year column is next up for a closer look. To look for any outliers or incorrect values, I wanted to pull up a quick view of the column statistics.`

In [40]:
# understand distribution of registration_year column
autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

`A min value of 1000 and max value of 9999 raises concerns, we're only living in 2020 after all (and if I remember my history lessons correctly, the car as we know it come into existence until the 19th century. Let's look at this more in depth:`

In [41]:
# freq dist of values in registration_year
autos['registration_year'].value_counts(ascending=False).sort_index()

1000    1
1001    1
1111    1
1800    2
1910    5
       ..
5911    1
6200    1
8888    1
9000    1
9999    3
Name: registration_year, Length: 95, dtype: int64

`Looks like on the head end of things, 1910 is most likely the earliest possible date that would be feasable under common sense. On the tail side, I want to expose a few more rows since the truncated results stopping at 5911 (a very futuristic car) suggests that there are more errenous values between the current day and 5911.` 

In [42]:
# freq dist of values in registration_year, tail end
autos['registration_year'].value_counts(ascending=False).sort_index().tail(15)

2015     392
2016    1220
2017    1392
2018     470
2019       2
2800       1
4100       1
4500       1
4800       1
5000       4
5911       1
6200       1
8888       1
9000       1
9999       3
Name: registration_year, dtype: int64

`If we look back to the ad creation dates, they are all from 2016. This analysis is being done in 2020, so the 2017/2018 dates seem like they could be feasable, but I don't know how registration years work in Germany so that would require a bit more analysis. To be conservative, I'm going to use 2017 as the cut off year on the high end, and 1910 as the cutoff year on the low end` 

In [43]:
# use boolean mask to update autos to only include rows where registration year is between 1909 and 2017
autos = autos[(autos['registration_year'] > 1909) & (autos['registration_year'] < 2017)]
                                                     

In [44]:
# display registration_year as a percentage of total after filtering to understand freq dist
autos['registration_year'].value_counts(normalize=True).sort_index()

1910    0.000107
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000043
          ...   
2012    0.028063
2013    0.017202
2014    0.014203
2015    0.008397
2016    0.026135
Name: registration_year, Length: 78, dtype: float64

## brand: a closer look 

In [45]:
#display names of unique values in brand column 
autos['brand'].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [46]:
#display freq dist of values in brand column
autos['brand'].value_counts().sort_index()

alfa_romeo         310
audi              4041
bmw               5137
chevrolet          266
chrysler           164
citroen            654
dacia              123
daewoo              70
daihatsu           117
fiat              1197
ford              3263
honda              366
hyundai            468
jaguar              73
jeep               106
kia                330
lada                27
lancia              50
land_rover          98
mazda              709
mercedes_benz     4503
mini               409
mitsubishi         384
nissan             713
opel              5022
peugeot           1393
porsche            286
renault           2201
rover               62
saab                77
seat               853
skoda              766
smart              661
sonstige_autos     458
subaru             100
suzuki             277
toyota             593
trabant             65
volkswagen        9862
volvo              427
Name: brand, dtype: int64

# brands to aggregate on 

I've chosen the following brands because they are the most popular ( > 1000 entries) based on value counts: 

* audi
* bmw
* ford 
* fiat 
* mercedes_benz
* opal
* peugeot
* renault 
* volkswagen 

## Aggregate mean price by brand 


In [47]:
# empty dict to accept for loop output on mean price by brand
brand_mean_prices = {}

# brands selected for analysis based on logic above
brands = ['audi', 'bmw', 'ford', 'fiat', 'mercedes_benz', 'opal', 'peugeot', 'renault', 'volkswagen']

# loop thru brands to extract mean price
for b in brands:
    selected_rows = autos[autos['brand'] == b]
    meanprice = selected_rows['price'].mean()
    brand_mean_prices[b] = meanprice

In [48]:
# display newly created mean price by brand dictionary
brand_mean_prices

{'audi': 9336.687453600594,
 'bmw': 8332.820517811953,
 'ford': 3749.4695065890287,
 'fiat': 2813.748538011696,
 'mercedes_benz': 8628.450366422385,
 'opal': nan,
 'peugeot': 3094.0172290021537,
 'renault': 2474.8646069968195,
 'volkswagen': 5402.410261610221}

### Aggregate mean price by mileage 

In [49]:
# empty dict to accept for loop output on mean price by mileage
brand_mean_mileage = {} 

# loop thru brands to extract mean mileage
for b in brands:
    selected_rows = autos[autos['brand'] == b]
    mean_mileage = selected_rows['odometer_km'].mean()
    brand_mean_mileage[b] = mean_mileage

In [50]:
# display newly created mean price by brand dictionary
brand_mean_mileage

{'audi': 129157.38678544914,
 'bmw': 132572.51313996495,
 'ford': 124266.01287159056,
 'fiat': 117121.9715956558,
 'mercedes_benz': 130788.36331334666,
 'opal': nan,
 'peugeot': 127153.62526920316,
 'renault': 128071.33121308497,
 'volkswagen': 128707.15879132022}

## Deeper Analysis: Storing Aggregate Data in DataFrame

`After noticing a price disparity between Audi, BMW, and Mercedes and the other brands selected, I wanted to see if there was a significant variation in the mean odometer mileage. To make the numbers easier to compare, I wanted to create a dataframe with mean price and mean odometer_km as column indicies for each brand name record index.`

In [51]:
# create mean price series
bmp_series = pd.Series(brand_mean_prices)

In [52]:
#create mean mileage series
bmm_series = pd.Series(brand_mean_mileage) 

In [53]:
# convert mean price series into df
df = pd.DataFrame(bmp_series, columns=['mean_price'])

In [54]:
# add mean mileage series as new column in mean price df
df['mean_mileage'] = bmm_series

In [55]:
# show newly creately df to ensure correct output 
print(df)

                mean_price   mean_mileage
audi           9336.687454  129157.386785
bmw            8332.820518  132572.513140
ford           3749.469507  124266.012872
fiat           2813.748538  117121.971596
mercedes_benz  8628.450366  130788.363313
opal                   NaN            NaN
peugeot        3094.017229  127153.625269
renault        2474.864607  128071.331213
volkswagen     5402.410262  128707.158791


`In this guided project, we practiced applying a variety of pandas methods to explore and understand a data set on car listings. Here are some next steps for you to consider:`

Data cleaning next steps:
* Identify categorical data that uses german words, translate them and map the values to their english counterparts
* Convert the dates to be uniform numeric data, so "2016-03-21" becomes the integer 20160321
* See if there are particular keywords in the name column that you can extract as new columns

Analysis next steps:
* Find the most common brand/model combinations
* Split the odometer_km into groups, and use aggregation to see if average prices follows any patterns based on the milage.
* How much cheaper are cars with damage than their non-damaged counterparts?

#### Nulls to Deal With

* vehicleType            44905 non-null object
* gearbox                47320 non-null object
* model                  47242 non-null object
* fuelType               45518 non-null object
* notRepairedDamage      40171 non-null object

---------------

#### axis=1 & dtype=int 

* yearOfRegistration     50000 non-null int64
* powerPS                50000 non-null int64
* monthOfRegistration    50000 non-null int64
* nrOfPictures           50000 non-null int64
* postalCode             50000 non-null int64

*remaining axis=1 dtype=object